### PR TITLE
Removed ROCM ifdef that governs thread count + smem parallel reduction.

### PR DESCRIPTION
**Pull request by 5had3z**
#149548 Fixed the arbitrarily missing parallelism for NLL, but they also added an arbritrary #ifdef ROCM guard around this fix to prevent its use on CUDA gpus. There is also a problem with the way the kernel does the reduction from the intermediate shared memory, using only thread 0 walking linearly. This has been changed to a simple parallel reduction algorithm.

Tested changes with `python3 test/test_nn.py` 

```
Ran 3551 tests in 200.554s

OK (skipped=998, expected failures=4)
```

Performance before and after with the script below with an RTX 3090, batch size x axis, time (sec) y axis. This GPU is also used for display graphics and such, so the measurements are pretty noisy, even with 100 samples.

## Before
before_nll

## After ifdef removal
after_nll

## After Parallel SMEM reduction

after_reduction

```python
import torch
from matplotlib import pyplot as plt
from torch.nn import functional as F

timing = []
batches=  list(range(32, 4096, 32))

for batch in [32] + batches:
    samples = []
    for _ in range(100):
        probs = torch.rand(batch, 10).cuda()
        labels = torch.randint(0, 10, (batch,)).cuda()
        start = torch.cuda.Event(enable_timing=True)
        end = torch.cuda.Event(enable_timing=True)
        start.record()
        F.nll_loss(probs, labels)
        end.record()
        torch.cuda.synchronize()
        elapsed = start.elapsed_time(end)
        samples.append(elapsed)
    timing.append(sum(samples) / len(samples))
timing = timing[1:]

plt.plot(batches, timing)
plt.show()
```

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd

### CODE REVIEW

**Comment by 5had3z** [comment]
Those failures don't seem to be related to my changes? The NN test also works fine when I call directly with `python3 test/test_nn.py TestNNDeviceTypeCUDA.test_variable_sequence_cuda_float32`, so maybe there is a race condition or something.

>TestNNDeviceTypeCUDA.test_variable_sequence_cuda_float32

>WIN: benchmark ('sum_floordiv_regression', 'compile_time_instruction_count') failed, actual result 971748003 is -5.29% lower than expected 1026000000 ±1.50% please update the expected results.  

I'm also curious as to the purpose of  `NLLLoss2d.cu`? torch.nn.NLLLoss2d is an alias for torch.nn.NLLLoss, so this seems to just be dead code?

**Review by jeffdaily — APPROVED** [review]

**Comment by jeffdaily** [comment]
@5had3z I'm attempting to rerun the failed CUDA UT job.  The CUDA benchmark job failure is unrelated.  Also, I forgot to trigger ROCm CI so I've done that now.  I approved your changes since they LGTM but pending CI passing.

**Comment by jeffdaily** [comment]
@5had3z rerunning job got the same error.  Let's try a rebase this time and see if we just happened to get an unlucky random set of inputs for that test.
